### PR TITLE
feat(platform/bitbucket): use development branching model as default branch

### LINKED ISSUE #20802: Bitbucket Cloud: initialize repos with development branch
### What would you like Renovate to be able to do?

Use development branch to initialize Bitbucket Cloud repositories.

Bitbucket Cloud supports a [variety of branching models](https://support.atlassian.com/bitbucket-cloud/docs/branch-a-repository/#The-branching-model), which are configured at the repository level.

Two of importance to this feature request are

- **Development branch** - this is where new feature branches are targeted, commonly `main` or `master` for trunk-based development and `develop` for git-flow.  This is configured under _Repository Details >> Worklow >> Branching Model_
- **Main branch** - this is configured under _Repository Details >> Advanced >> Main branch_.  Commonly this is left set to `main` or `master` regardless of trunk-based, git-flow or other workflow

At present, when Bitbucket Cloud repositories are initialized, it sets the defaultBranch to be the repositories **main branch** instead of the preferred **development branch**.  This means that for repositories using non trunk-based workflows, Renovate PRs will be targeting the wrong branch for their workflow.

This same situation was discussed on https://github.com/renovatebot/renovate/discussions/11739 a few years back with a suggested workaround of using the [baseBranches](https://docs.renovatebot.com/configuration-options/#basebranches).  

This feature request would remove the need for such a workaround.

### If you have any ideas on how this should be implemented, please tell us here.

Update the Bitbucket repositories initialization to fetch the **development branch** from the [REST API: Branching Model for the Repository](https://developer.atlassian.com/cloud/bitbucket/rest/api-group-branching-model/#api-repositories-workspace-repo-slug-branching-model-get) and set this as the Renovate repository default branch.

### Is this a feature you are interested in implementing yourself?

Yes

### PR DESCRIPTION
**Pull request by setchy**
## Changes

Updates the Bitbucket Cloud repository initialization logic to fetch and set the development branch name instead of relying only on the main branch name.

## Context

- Closes #20802

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [x] Both unit tests + ran on a real repository

### CODE REVIEW

**Review by rarkins — CHANGES_REQUESTED** [review]
Please mark as breaking and target v35 branch

**Comment by setchy** [comment]
> Please mark as breaking and target v35 branch

Sure - can do.  

Will add a few more test cases also

**Comment by rarkins** [comment]
Blocked by #20800

**Review by rarkins — CHANGES_REQUESTED** [review]
Needs docs

File: lib/modules/platform/bitbucket/index.ts [hunk 1]
```diff
@@ -179,7 +181,17 @@ export async function initRepo({
...
+    // Fetch Bitbucket development branch
+    developmentBranch = (
+      await bitbucketHttp.getJson<RepoBranchingModel>(
+        `/2.0/repositories/${repository}/branching-model`
+      )
+    ).body.development?.branch?.name;
+
+    config.defaultBranch = developmentBranch
+      ? developmentBranch
+      : info.mainBranch;
```
Comment (viceice) [inline lib/modules/platform/bitbucket/index.ts:194]:
```suggestion
    config.defaultBranch = developmentBranch
      ?? info.mainBranch;
```

**Review by viceice — CHANGES_REQUESTED** [review]
this is a breaking change, we should disable new behavior with a repo config option.

**Comment by setchy** [comment]
Added `bbUseDevelopmentBranch` config option (default to false), docs and updated tests

File: docs/usage/configuration-options.md [hunk 2]
```diff
@@ -276,6 +276,14 @@ With a negation, all branches except those matching the regex will be added to t
 
 Configuring this to `true` means that Renovate will detect and apply the default reviewers rules to PRs (Bitbucket only).
 
+## bbUseDevelopmentBranch
+
+By default, Renovate will use the repositories `main branch`.
```
Comment (rarkins) [inline docs/usage/configuration-options.md:281]:
```suggestion
By default, Renovate will use a repository's default branch (typically called `main` or `master`).
```

This is same terminology used by Atlassian here:

image

File: docs/usage/configuration-options.md [hunk 3]
```diff
@@ -276,6 +276,14 @@ With a negation, all branches except those matching the regex will be added to t
...
+
+Configuring this to `true` means that Renovate will detect and use the Bitbucket [development branch](https://support.atlassian.com/bitbucket-cloud/docs/branch-a-repository/#The-branching-model) as defined by the repositories branching model.
```
Comment (rarkins) [inline docs/usage/configuration-options.md:283]:
```suggestion
Configuring this to `true` means that Renovate will detect and use the Bitbucket [development branch](https://support.atlassian.com/bitbucket-cloud/docs/branch-a-repository/#The-branching-model) as defined by the repository's branching model.
```

File: docs/usage/configuration-options.md [hunk 4]
```diff
@@ -276,6 +276,14 @@ With a negation, all branches except those matching the regex will be added to t
...
+
+If the `development branch` is configured but the branch itself does not exist (ie: it was deleted), Renovate will fall back to using the repositories `main branch`.
```
Comment (rarkins) [inline docs/usage/configuration-options.md:285]:
```suggestion
If the "development branch" is configured but the branch itself does not exist (e.g. it was deleted), Renovate will fall back to using the repository's default branch.
```

File: lib/config/options/index.ts [hunk 5]
```diff
@@ -1623,6 +1623,14 @@ const options: RenovateOptions[] = [
     default: true,
     supportedPlatforms: ['bitbucket', 'bitbucket-server'],
   },
+  {
+    name: 'bbUseDevelopmentBranch',
+    description:
+      'Use the [development branch](https://support.atlassian.com/bitbucket-cloud/docs/branch-a-repository/#The-branching-model) as the default target branch for pull requests (Bitbucket only).',
```
Comment (rarkins) [inline lib/config/options/index.ts:1629]:
```suggestion
      'Use the repository's [development branch](https://support.atlassian.com/bitbucket-cloud/docs/branch-a-repository/#The-branching-model) as the repository's default branch.',
```

File: lib/config/options/index.ts [hunk 6]
```diff
@@ -1623,6 +1623,14 @@ const options: RenovateOptions[] = [
     default: true,
     supportedPlatforms: ['bitbucket', 'bitbucket-server'],
   },
+  {
+    name: 'bbUseDevelopmentBranch',
```
Comment (rarkins) [inline lib/config/options/index.ts:1627]: I think this needs to be `globalOnly` because by the time config is read from the repo, it would be too late.

File: lib/modules/platform/bitbucket/index.ts [hunk 7]
```diff
@@ -179,7 +183,21 @@ export async function initRepo({
...
+          `/2.0/repositories/${repository}/branching-model`
+        )
+      ).body.development?.branch?.name;
+
+      mainBranch = developmentBranch ?? info.mainBranch;
+    } else {
+      mainBranch = info.mainBranch;
+    }
+
+    config.defaultBranch = mainBranch;
```
Comment (rarkins) [inline lib/modules/platform/bitbucket/index.ts:200]:
```suggestion
    config.defaultBranch = mainBranch;

    if (bbUseDevelopmentBranch) {
      // Fetch Bitbucket development branch
      developmentBranch = (
        await bitbucketHttp.getJson<RepoBranchingModel>(
          `/2.0/repositories/${repository}/branching-model`
        )
      ).body.development?.branch?.name;

      if (developmentBranch) {
        config.defaultBranch = developmentBranch;
      }
    }

```

I don't see where you are checking if it exists though, as described in the new docs?

**Review by rarkins — CHANGES_REQUESTED** [review]

File: lib/modules/platform/bitbucket/index.ts (hunk 7 above)
Comment (setchy) [reply lib/modules/platform/bitbucket/index.ts:200]: It's part of the returned object from calling[ Repository Branching Model API](https://developer.atlassian.com/cloud/bitbucket/rest/api-group-branching-model/#api-repositories-workspace-repo-slug-branching-model-get)

> Always has a development property. development.branch contains the actual repository branch object that is considered to be the development branch. development.branch will not be present if it does not exist.


The response is guaranteed to always return a `development.name`, such as
```
{
    ...
    "development": {
        "name": "main"
    }
    ...
}
```
or 
```
{
    ...
    "development": {
        "name": "develop"
    }
    ...
}
```
or whatever the repository admin has specified....

But it will only return a `development.branch.name` if that branch exists....
eg:
```
{
    ...
    "development": {
        "name": "develop",
         "use_mainbranch": false,
          "branch": {
              "name": "develop",
              "target": {
                  "hash": "e6aded4501c954c435174f399492fb1a854be188"
              },
              "type": "branch"
          }
    }
    ...
}
 ```

File: docs/usage/configuration-options.md (hunk 2 above)
Comment (setchy) [reply docs/usage/configuration-options.md:281]: Yeah, i noticed late last night that their documentation language uses `default branch` but that isn't consistent with what their product uses.  

Attached are some screenshots showing what they use across their Advanced configuration, branching model and branches screens

**Advanced configuration for repository**
Screenshot 2023-03-08 at 8 57 51 AM

**main-based development branches**
Screenshot 2023-03-08 at 8 58 13 AM

**git-flow based development branches**
Screenshot 2023-03-08 at 8 58 26 AM

**branching model configuration**
Screenshot 2023-03-08 at 9 00 47 AM

Comment (setchy) [reply docs/usage/configuration-options.md:281]: Happy to update the docs to the proposed wording if you still think that makes most sense

File: lib/modules/platform/bitbucket/index.ts (hunk 7 above)
Comment (rarkins) [reply lib/modules/platform/bitbucket/index.ts:200]: Gotcha

File: docs/usage/configuration-options.md (hunk 2 above)
Comment (rarkins) [reply docs/usage/configuration-options.md:281]: Hard to tell what to do if they are inconsistent themselves. Seems like main is used more

**Comment by setchy** [comment]
> Blocked by #20800

I see #20800 and #20811 as being related to this change, but not blocking this change.  

I'd like to propose this feature request moves forward as it has benefits for any new PRs and there are a few manual workarounds teams can use (or currently use) to correct existing PRs, such as
- manually updating the open PR to target their `develop` branch
- rename the Bitbucket PR, close it and then let Renovate raise a new one with the correct target branch during the next cycle

File: lib/config/options/index.ts (hunk 6 above)
Comment (setchy) [reply lib/config/options/index.ts:1627]: The mainBranch / developmentBranch logic is currently within the `initRepo` phase.  

Wouldn't this mean that `bbUseDevelopmentBranch` can be used as either a global config or a repository config?

**Comment by setchy** [comment]
> 

avoided the breaking change with the new `bbUseDevelopmentBranch` config options which is defaulted to false.

@rarkins - with this in place, is it still OK to target this PR to `v35`?

File: lib/config/options/index.ts (hunk 6 above)
Comment (rarkins) [reply lib/config/options/index.ts:1627]: initRepo is done prior to reading config

**Comment by viceice** [comment]
> > this is a breaking change, we should disable new behavior with a repo config option.
> 
> was able to make this a non-breaking change by adding the `bbUseDevelopmentBranch` config option and defaulting it to false.
> 
> @rarkins - with this in place, is it still OK to target this PR to `v35`?

you can target main when it's not breaking.

do you plan to support more branching models? so this config option should be a enum with null being disabled / current behavior ?

File: lib/modules/platform/bitbucket/types.ts [hunk 8]
```diff
@@ -30,12 +30,21 @@ export interface PagedResult<T = any> {
 export interface RepoInfo {
   isFork: boolean;
   owner: string;
-  mainbranch: string;
+  mainBranch: string;
```
Comment (viceice) [inline lib/modules/platform/bitbucket/types.ts:33]: please revert this unrelated change.

File: lib/config/options/index.ts [hunk 9]
```diff
@@ -1623,6 +1623,14 @@ const options: RenovateOptions[] = [
...
+    description: `Use the repository's [development branch](https://support.atlassian.com/bitbucket-cloud/docs/branch-a-repository/#The-branching-model) as the repository's default branch.`,
+    type: 'boolean',
+    default: false,
+    supportedPlatforms: ['bitbucket'],
+    globalOnly: true,
```
[the remark this answers is not in the document]
Comment (setchy) [reply lib/config/options/index.ts:1632]: See [this](https://github.com/renovatebot/renovate/pull/20803#discussion_r1129434027) conversation

**Comment by setchy** [comment]
> do you plan to support more branching models? so this config option should be a enum with null being disabled / current behavior ?

Can you please explain a bit more about what you mean by "more branching models"?

**Development branch** is Atlassian Bitbucket terminology and can be configured to any branch. Examples below on how this could be configured by each repository admin

Screenshot 2023-03-09 at 2 21 29 PM
Screenshot 2023-03-09 at 2 23 25 PM
Screenshot 2023-03-09 at 2 23 07 PM
Screenshot 2023-03-09 at 2 24 25 PM

File: lib/modules/platform/bitbucket/types.ts (hunk 8 above)
Comment (setchy) [reply lib/modules/platform/bitbucket/types.ts:33]: sure... i'll just learn to look past the camelCase mismatch 🙈

File: docs/usage/self-hosted-configuration.md [hunk 10]
```diff
@@ -139,6 +139,14 @@ For example:
 }
 ```
 
+## bbUseDevelopmentBranch
+
+By default, Renovate will use a repository's "main branch" (typically called `main` or `master`) as the "default branch".
+
+Configuring this to `true` means that Renovate will detect and use the Bitbucket [development branch](https://support.atlassian.com/bitbucket-cloud/docs/branch-a-repository/#The-branching-model) as defined by the repository's branching model.
+
+If the "development branch" is configured but the branch itself does not exist (e.g. it was deleted), Renovate will fall back to using the repository's "main branch".
```
[the remark this answers is not in the document]
Comment (setchy) [reply docs/usage/self-hosted-configuration.md:148]: This fall-back matched the behavior via Bitbucket Cloud UI.  Attached is proof of where I simulated
- repository branching model specifies a specific "development branch" called `whatever-i-like` 
- delete "development branch" `whatever-i-like`
- create a PR.  UI will fall back to the "main branch" as defined in the repositories advanced configuration.  "main branches" cannot be deleted
Screenshot 2023-03-09 at 2 40 46 PM
Screenshot 2023-03-09 at 2 41 01 PM
Screenshot 2023-03-09 at 2 41 12 PM

File: lib/modules/platform/bitbucket/index.ts [hunk 11]
```diff
@@ -179,7 +183,23 @@ export async function initRepo({
...
+    if (bbUseDevelopmentBranch) {
+      // Fetch Bitbucket development branch
+      developmentBranch = (
+        await bitbucketHttp.getJson<RepoBranchingModel>(
+          `/2.0/repositories/${repository}/branching-model`
+        )
+      ).body.development?.branch?.name;
+
+      if (developmentBranch) {
+        mainBranch = developmentBranch;
```
Comment (viceice) [inline lib/modules/platform/bitbucket/index.ts:198]: this also means that the repo config is read from development branch. is that expected?

Comment (setchy) [reply lib/modules/platform/bitbucket/index.ts:198]: Hmm, good call out.  

I just had a re-read of [this gitflow article](https://www.atlassian.com/git/tutorials/comparing-workflows/gitflow-workflow) and it would seem to suggest that continuing to read the repo config from the "main branch" would make sense.

Is that what you takeaway too @viceice?  @rarkins how about you?

File: docs/usage/self-hosted-configuration.md (hunk 10 above)
[the remark this answers is not in the document]
Comment (rarkins) [reply docs/usage/self-hosted-configuration.md:148]: Yes, seems that bitbucket gracefully continues so let's do the same

File: lib/modules/platform/bitbucket/index.ts (hunk 11 above)
Comment (rarkins) [reply lib/modules/platform/bitbucket/index.ts:198]: I had simply assumed that it was your intent to read config from the development branch, so thanks @viceice for calling it out.

In that case it's a bit more complicated!

So we could, as this PR has done, use the development branch as _default_ branch in Renovate and read config + dependencies from there (assuming `baseBranches` is not configured and overrides that).

Alternatively, our behavior could be that `baseBranches=[developmentBranch]` unless `baseBranches` are explicitly configured by the user. In other words default branch, which config is read from, remains as-is. This explains why @viceice thought that this new option could be read from repo config while I thought it could not.

I'm ok with the second option

Comment (setchy) [reply lib/modules/platform/bitbucket/index.ts:198]: Having slept on it - I'm leaning back towards keeping the implementation as-is and reading the configuration from the `develop` branch.  

Why?  this is the most active branch used for "integrating" feature branches.  it will be the one where developers are committing to most often.  the "main branch" is really just used for production releases.  If we were to read the renovate config from "main branch", that would mean teams that want to change their renovate configuration would have to do through a feature branch -> develop branch -> prod release workflow (assuming they dont commit directly to main 🤠 )

**Comment by rarkins** [comment]
Sorry this was closed unintentionally by GitHub when v35 was closed

**Comment by setchy** [comment]
> Sorry this was closed unintentionally by GitHub when v35 was closed

All good.  Will re-open a new PR to continue the discussion
